### PR TITLE
Feat: refactore `EIP-4844`

### PR DESCRIPTION
**Pull request by mrLSD**
## Description

📶  This PR related to CANCUN hard for PR: #39 

➡️ Refactoring `EIP-4844` logic - blob transactions
➡️ Fix `ethereum-json-tests` for `EIP-4844`

### CODE REVIEW

File: evm-tests/ethjson/src/test_helpers/state.rs [hunk 1]
```diff
@@ -79,30 +76,12 @@ pub struct MultiTransaction {
 
 	/// EIP-4844
 	#[serde(default)]
-	pub blob_versioned_hashes: Vec<H256>,
+	pub blob_versioned_hashes: Vec<U256>,
```
Comment (aleksuss) [inline evm-tests/ethjson/src/test_helpers/state.rs:79]: Just out of curiosity. Why did you change `H256` to `U256` here? As I remember, the `H256` type is used for hash storing.

File: evm-tests/jsontests/src/state.rs [hunk 2]
```diff
@@ -124,14 +116,18 @@ impl Test {
...
-				u256_to_h256(r.0)
+				crate::utils::u256_to_h256(r.0)
 			})
 		} else {
 			None
 		};
 
-		Some(MemoryVicinity {
+		#[allow(clippy::map_clone)]
+		let blob_hashes = tx.blob_versioned_hashes.iter().map(|v| *v).collect();
```
Comment (aleksuss) [inline evm-tests/jsontests/src/state.rs:126]: I didn't get it. Why not just clone the `tx.blob_versioned_hashes`?

File: evm-tests/jsontests/src/state.rs [hunk 3]
```diff
@@ -426,22 +426,185 @@ pub fn test(
...
+				InvalidTxReason::OutOfFund => {
+					let check_result = exception
+						== "TransactionException.INSUFFICIENT_ACCOUNT_FUNDS"
+						|| exception == "TR_TypeNotSupported"
+						|| exception == "TR_NoFunds"
+						|| exception == "TR_NoFundsX"
+						|| exception == "TransactionException.INSUFFICIENT_MAX_FEE_PER_BLOB_GAS";
+					assert!(
+						check_result,
+						"unexpected exception {exception:?} for OutOfFund  for test: {name}"
```
Comment (aleksuss) [inline evm-tests/jsontests/src/state.rs:521]: Redundant space and below as well.
```suggestion
						"unexpected exception {exception:?} for OutOfFund for test: {name}"
```

File: evm-tests/jsontests/src/state.rs [hunk 4]
```diff
@@ -450,7 +613,11 @@ fn test_run(
 	specific_spec: Option<ForkSpec>,
 ) -> TestExecutionResult {
 	let mut tests_result = TestExecutionResult::new();
+	let test_tx = &test.0.transaction;
 	for (spec, states) in &test.0.post_states {
+		// if name != "create2collisionSelfdestructedOOG" {
```
Comment (aleksuss) [inline evm-tests/jsontests/src/state.rs:618]: Should be removed?

File: evm-tests/jsontests/src/state.rs [hunk 5]
```diff
@@ -593,30 +830,43 @@ fn test_run(
...
+					.state_mut()
+					.deposit(caller, amount_to_return_for_caller);
 
 				let (values, logs) = executor.into_state().deconstruct();
 
 				backend.apply(values, logs, delete_empty);
+			} else {
+				// TODOFEE
+				if let Some(_e) = state.expect_exception.as_ref() {
+					// println!("==> expect_exception: {e}");
```
Comment (aleksuss) [inline evm-tests/jsontests/src/state.rs:863]: If you don't need this `println!` anymore then it could be rewritten.

File: evm-tests/jsontests/tests/state.rs [hunk 6]
```diff
@@ -77,6 +75,7 @@ const SKIPPED_CASES: &[&str] = &[
 ];
 
 fn should_skip(path: &Path) -> bool {
+	println!("{path:?}");
```
Comment (aleksuss) [inline evm-tests/jsontests/tests/state.rs:78]: Should be removed?

**Comment by mrLSD** [comment]
@aleksuss it was an issue related to merging `feat/evm-tests` - some changes was not applied. Fixed it.

File: evm-tests/jsontests/src/state.rs [hunk 7]
```diff
@@ -426,22 +424,185 @@ pub fn test(
...
+						check_result,
+						"unexpected exception {exception:?} for GasLimitReached for test: {name}"
+					);
+				}
+				InvalidTxReason::IntrinsicGas => {
+					let check_result = exception == "TR_NoFundsOrGas"
+						|| exception == "TR_IntrinsicGas"
+						|| exception == "TransactionException.INTRINSIC_GAS_TOO_LOW"
+						|| exception == "IntrinsicGas"
+						|| exception == "TR_TypeNotSupported";
```
Comment (birchmd) [inline evm-tests/jsontests/src/state.rs:534]: Why is `TR_TypeNotSupported` lumped in with the intrinsic gas errors?

**Review by birchmd — APPROVED** [review]
Nice job, thanks for all your work on this!

File: evm-tests/jsontests/src/state.rs (hunk 7 above)
Comment (mrLSD) [reply evm-tests/jsontests/src/state.rs:534]: I don't know. It's only one case for `Berlin` hardfork.

File: evm-tests/jsontests/src/main.rs [hunk 8]
```diff
@@ -188,29 +188,58 @@ const SKIPPED_CASES: &[&str] = &[
...
-	"vmPerformance/loopMul",
-	"stTimeConsuming/CALLBlake2f_MaxRounds",
-	// Skip python-specific tests
-	"eip4844_blobs",
-	// Cancun blob txs
-	"stEIP4844-blobtransactions",
+	// These tests are passing, but they take a lot of time to execute so can going to skip them.
+	// "stTimeConsuming/static_Call50000_sha256",
+	// "vmPerformance/loopMul",
+	// "stTimeConsuming/CALLBlake2f_MaxRounds",
```
[the remark this answers is not in the document]
Comment (mrLSD) [reply evm-tests/jsontests/src/main.rs:194]: Actually, I left it to know what kind of tests is slow. If enabling it, tests can pass faster about 15 min. It's really slow tests.

[the remark this answers is not in the document]
Comment (mrLSD) [reply evm-tests/jsontests/src/main.rs:194]: It's good for CI, but for real life development it defenetely should be enabled (uncommented).

[the remark this answers is not in the document]
Comment (birchmd) [reply evm-tests/jsontests/src/main.rs]: You could maybe use a compile-time feature flag which is disabled by default, but enabled for CI.

[the remark this answers is not in the document]
Comment (aleksuss) [reply evm-tests/jsontests/src/main.rs]: I like the idea proposed by @birchmd about enabling it for the CI only. @mrLSD, what do you think about that?

[the remark this answers is not in the document]
Comment (mrLSD) [reply evm-tests/jsontests/src/main.rs]: Managed it here: #43

**Review by aleksuss — APPROVED** [review]
